### PR TITLE
[FIX] Send: better show input errors (RT-2975)

### PR DESCRIPTION
**Pull request by darkdarkdragon**
Re-validate input when currency changed,
and do not start pathfinding if input
is not valid

### CODE REVIEW

**Comment by jublo** [comment]
Some unit tests failed, please take a look here: https://travis-ci.org/ripple/ripple-client/builds/47371201#L969, and possibly fix them.

**Comment by darkdarkdragon** [comment]
@jublo fixed

File: test/unit/tabs/exchangeTabSpec.js [hunk 1]
```diff
@@ -45,9 +46,13 @@ describe('ExchangeCtrl', function() {
         account: 'r4EwBWxrx5HxYRyisfGzMto3AT8FZiYdWk'
       }
     };
+    */
+    rpId.loginStatus = true;
+    rpId.account = 'r4EwBWxrx5HxYRyisfGzMto3AT8FZiYdWk';
 
-
-    ctrl = $controller("ExchangeCtrl", dependencies);
+    var templ = angular.element( rippleclient.tabs.exchange.generateHtml() );
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: no spaces inside brackets

File: test/unit/tabs/exchangeTabSpec.js [hunk 2]
```diff
@@ -45,9 +46,13 @@ describe('ExchangeCtrl', function() {
...
+    var element = $compile( templ )(scope);
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: see above

File: test/unit/tabs/exchangeTabSpec.js [hunk 3]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
     done();
   });
 
+
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: double newlines are not needed imo

File: test/unit/tabs/exchangeTabSpec.js [hunk 4]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
...
     var spy = sinon.spy(scope, 'reset_paths');
     scope.$apply();
 
-    assert(spy.called);
-    assert.strictEqual(scope.exchange.alternatives.length, 0);
-    assert.strictEqual(scope.exchange.amount_feedback.to_text_full(), "2/XRP");
+    setTimeout(function() {
+      assert(spy.called, 'reset_paths not called');
+      assert.strictEqual(scope.exchange.alternatives.length, 0, 'paths not empty');
+      assert.strictEqual(scope.exchange.amount_feedback && scope.exchange.amount_feedback.to_text_full(), "2/XRP", 'name wrong');
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: `'2/XRP'` with single quotes

File: test/unit/tabs/exchangeTabSpec.js [hunk 5]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
...
-    done();
+      done();
+    }, 100);
   });
 
+
   it('should update paths after changing currency choice', function(done) {
 
     var spy = sinon.spy(scope, 'update_exchange');
+    scope.exchange.amount = "2";
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: single quotes

File: test/unit/tabs/exchangeTabSpec.js [hunk 6]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
...
     scope.exchange.currency_name = "USD";
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: see above

File: test/unit/tabs/exchangeTabSpec.js [hunk 7]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
...
 
-    assert.isObject(scope.exchange.currency_obj);
-    assert.strictEqual(scope.exchange.currency_code, 'USD');
-    assert.strictEqual(scope.exchange.currency_name, 'USD - US Dollar');
+      assert.isObject(scope.exchange.currency_obj, 'currency_obj not object');
+      assert.strictEqual(scope.exchange.currency_code, 'USD', 'currency_code not USD');
+      assert.strictEqual(scope.exchange.currency_name, 'USD - US Dollar', 'currency_name not USD');
 
-    assert.strictEqual(scope.exchange.path_status, 'waiting');
+      assert.strictEqual(scope.exchange.path_status, 'pending', 'path_status to pending');
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: Thanks for having fixed the indention! :+1:

File: test/unit/tabs/exchangeTabSpec.js [hunk 8]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
...
 
-    done();
+      done();
+    }, 100);
   });
 
-
   it('should update paths after entering 2 as amount for XRP', function(done) {
 
     scope.exchange.amount = "0.001";
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: While you're at it.

File: test/unit/tabs/exchangeTabSpec.js [hunk 9]
```diff
@@ -126,13 +138,13 @@ describe('ExchangeCtrl', function() {
     var spy = sinon.spy(scope, 'reset_paths');
     scope.$apply();
 
-    assert(spy.called);
-    assert.strictEqual(scope.exchange.alternatives.length, 0);
-    assert.strictEqual(scope.exchange.amount_feedback.to_text_full(), "0.001/USD/r4EwBWxrx5HxYRyisfGzMto3AT8FZiYdWk");
+    setTimeout(function() {
+      assert(spy.called, 'reset_paths not called for USD');
+      assert.strictEqual(scope.exchange.alternatives.length, 0, 'paths not null');
+      assert.strictEqual(scope.exchange.amount_feedback.to_text_full(), "0.001/USD/r4EwBWxrx5HxYRyisfGzMto3AT8FZiYdWk", 'feedback invalid');
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: same here

File: test/unit/tabs/exchangeTabSpec.js [hunk 10]
```diff
@@ -126,13 +138,13 @@ describe('ExchangeCtrl', function() {
...
-    done();
+      done();
+    }, 100);
   });
 
-
-
-});
\ No newline at end of file
+});
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: Thanks for having added the trailing newline! :star2:

File: test/unit/tabs/exchangeTabSpec.js [hunk 11]
```diff
@@ -87,38 +92,45 @@ describe('ExchangeCtrl', function() {
     done();
   });
 
+  
   it('should update paths after entering 2 as amount for XRP', function(done) {
 
     scope.exchange.amount = "2";
```
Comment (jublo) [inline test/unit/tabs/exchangeTabSpec.js]: Got some time for these quote signs, too?

**Comment by darkdarkdragon** [comment]
@jublo fixed

**Comment by mrajvanshy** [comment]
LGTM

**Pull request closed without merging**